### PR TITLE
feat: added trigger's sink failures count metric

### LINKED ISSUE #465: Expose Trigger's sink failures count metric
**Problem**

A metric to easily see that a Trigger isn't working, because the trigger's sink is failing (e.g. it always returns a 4xx/5xx or times out, DNS lookup fails, etc.).

**[Persona:](https://github.com/knative/eventing/blob/master/docs/personas.md)** 
Which persona is this feature for?
Event Consumer

**Exit Criteria**

Expose Trigger's sink failures count metric.

**Time Estimate (optional):** 
How many developer-days do you think this may take to resolve?
1

**Additional context (optional)**

We can expose the following metric:

- `sink_failures_total{id=<trigger_uid>, url=<subscriber_url>} <num_failures>`

  - The `id` label allows correlating this metric with Kafka consumer metrics like:
    `kafka_consumer_request_rate{client_id="consumer-<trigger_uid>-1",kafka_version="2.6.0",} <value>`
    `kafka_consumer_io_wait_ratio{client_id="consumer-<trigger_uid>-1",kafka_version="2.6.0",} <value>`
    `...`
  - The `url` label allows correlating failures of a specific sink since we can have 2 triggers with the same sink URL.

### PR DESCRIPTION
**Pull request by Abhishek357**
Fixes #465

## Proposed Changes

- Added a new counter metric: `event_dispatch_retries` that counts how many times we retry a dispatch operation.

### CODE REVIEW

File: data-plane/core/src/main/java/dev/knative/eventing/kafka/broker/core/metrics/Metrics.java [hunk 1]
```diff
@@ -106,6 +106,12 @@ public class Metrics {
    */
   public static final String EVENT_DISPATCH_LATENCY = "event_dispatch_latencies";
 
+  /**
+   * @link https://knative.dev/docs/eventing/observability/metrics/eventing-metrics/
+   * @see Metrics#eventDispatchRetry(io.micrometer.core.instrument.Tags)
+   */
+  public static final String EVENT_DISPATCH_RETRY = "event_dispatch_retries";
```
Comment (pierDipi) [inline data-plane/core/src/main/java/dev/knative/eventing/kafka/broker/core/metrics/Metrics.java:113]:
```suggestion
  public static final String EVENT_DISPATCH_RETRY = "event_dispatch_retries_count";
```

File: data-plane/core/src/main/java/dev/knative/eventing/kafka/broker/core/metrics/Metrics.java [hunk 2]
```diff
@@ -291,80 +297,87 @@ private static String getHost() {
...
-      .baseUnit(BaseUnits.MILLISECONDS)
-      .serviceLevelObjectives(LATENCY_SLOs);
+        .builder(EVENT_PROCESSING_LATENCY)
+        .description("The time spent processing an event")
+        .tags(tags)
+        .baseUnit(BaseUnits.MILLISECONDS)
+        .serviceLevelObjectives(LATENCY_SLOs);
+  }
+
+  public static Counter.Builder eventDispatchRetry(final io.micrometer.core.instrument.Tags tags) {
```
Comment (pierDipi) [inline data-plane/core/src/main/java/dev/knative/eventing/kafka/broker/core/metrics/Metrics.java:338]:
```suggestion
  public static Counter.Builder eventDispatchRetriesCount(final io.micrometer.core.instrument.Tags tags) {
```

File: data-plane/dispatcher/src/main/java/dev/knative/eventing/kafka/broker/dispatcher/impl/http/WebClientCloudEventSender.java [hunk 3]
```diff
@@ -57,6 +63,9 @@ public final class WebClientCloudEventSender implements CloudEventSender {
   private final AtomicBoolean closed = new AtomicBoolean(false);
   private final AtomicInteger inFlightRequests = new AtomicInteger(0);
 
+  private final MeterRegistry meterRegistry;
+  private final Tags noResponseResourceTags;
```
Comment (pierDipi) [inline data-plane/dispatcher/src/main/java/dev/knative/eventing/kafka/broker/dispatcher/impl/http/WebClientCloudEventSender.java:67]: We should initialize these 2 fields in the constructor

File: data-plane/dispatcher/src/main/java/dev/knative/eventing/kafka/broker/dispatcher/impl/http/WebClientCloudEventSender.java [hunk 4]
```diff
@@ -130,13 +139,36 @@ private Future<HttpResponse<Buffer>> send(final CloudEvent event, final int retr
...
           return Future.failedFuture(cause);
         }
 
         if (retryCounter < consumerVerticleContext.getEgressConfig().getRetry()) {
+
+          Metrics
+          .eventCount(this.consumerVerticleContext.getTags())
+          .register(meterRegistry)
+          .increment();          
+
```
Comment (pierDipi) [inline data-plane/dispatcher/src/main/java/dev/knative/eventing/kafka/broker/dispatcher/impl/http/WebClientCloudEventSender.java:171]: We can do the metric increment as part of the `retry` method directly, something like this:

```java
  private Future<HttpResponse<Buffer>> retry(int retryCounter, HttpResponse response, CloudEvent event) {
    final var tags = getTags(response);  
    Metrics
      .eventDispatchRetriesCount(tags)
      .register(meterRegistry)
      .increment();

    Promise<HttpResponse<Buffer>> r = Promise.promise();
    final var delay = retryPolicyFunc.apply(retryCounter + 1);
    vertx.setTimer(delay, v -> send(event, retryCounter + 1).onComplete(r));
    return r.future();
  }
```

in this way, we remove the duplication

File: data-plane/core/src/main/java/dev/knative/eventing/kafka/broker/core/metrics/Metrics.java
Comment (pierDipi) [inline data-plane/core/src/main/java/dev/knative/eventing/kafka/broker/core/metrics/Metrics.java:1]: Formatting/indentation changes in this file should be reverted unless they are related to the PR itself

File: data-plane/core/src/main/java/dev/knative/eventing/kafka/broker/core/metrics/Metrics.java (hunk 1 above)
Comment (Abhishek357) [reply data-plane/core/src/main/java/dev/knative/eventing/kafka/broker/core/metrics/Metrics.java:113]: Yes @matzew, we can name it as `event_dispatch_retry_count` @pierDipi WDYT?

Comment (pierDipi) [reply data-plane/core/src/main/java/dev/knative/eventing/kafka/broker/core/metrics/Metrics.java:113]: Sounds good

Comment (pierDipi) [reply data-plane/core/src/main/java/dev/knative/eventing/kafka/broker/core/metrics/Metrics.java:113]: Let's align the method name as well

Comment (Abhishek357) [reply data-plane/core/src/main/java/dev/knative/eventing/kafka/broker/core/metrics/Metrics.java]: done

File: data-plane/core/src/main/java/dev/knative/eventing/kafka/broker/core/metrics/Metrics.java (hunk 2 above)
Comment (Abhishek357) [reply data-plane/core/src/main/java/dev/knative/eventing/kafka/broker/core/metrics/Metrics.java]: addressed

File: data-plane/dispatcher/src/main/java/dev/knative/eventing/kafka/broker/dispatcher/impl/http/WebClientCloudEventSender.java (hunk 3 above)
Comment (Abhishek357) [reply data-plane/dispatcher/src/main/java/dev/knative/eventing/kafka/broker/dispatcher/impl/http/WebClientCloudEventSender.java:69]: addressed

File: data-plane/dispatcher/src/main/java/dev/knative/eventing/kafka/broker/dispatcher/impl/http/WebClientCloudEventSender.java (hunk 4 above)
Comment (Abhishek357) [reply data-plane/dispatcher/src/main/java/dev/knative/eventing/kafka/broker/dispatcher/impl/http/WebClientCloudEventSender.java]: addressed

File: data-plane/core/src/main/java/dev/knative/eventing/kafka/broker/core/metrics/Metrics.java
Comment (Abhishek357) [reply data-plane/core/src/main/java/dev/knative/eventing/kafka/broker/core/metrics/Metrics.java:1]: addressed

File: data-plane/dispatcher/src/main/java/dev/knative/eventing/kafka/broker/dispatcher/impl/http/WebClientCloudEventSender.java [hunk 5]
```diff
@@ -60,6 +65,10 @@ public final class WebClientCloudEventSender implements CloudEventSender {
   private final AtomicBoolean closed = new AtomicBoolean(false);
   private final AtomicInteger inFlightRequests = new AtomicInteger(0);
 
+  private final MeterRegistry meterRegistry;
+  private final Tags noResponseResourceTags;
+  
+  private static final Tag NO_RESPONSE_CODE_CLASS_TAG = Tag.of(Metrics.Tags.RESPONSE_CODE_CLASS, "5xx");
```
Comment (pierDipi) [inline data-plane/dispatcher/src/main/java/dev/knative/eventing/kafka/broker/dispatcher/impl/http/WebClientCloudEventSender.java:71]: Is this tag defined somewhere else?

File: data-plane/dispatcher/src/main/java/dev/knative/eventing/kafka/broker/dispatcher/impl/http/WebClientCloudEventSender.java [hunk 6]
```diff
@@ -139,22 +151,36 @@ private Future<HttpResponse<Buffer>> send(final CloudEvent event, final int retr
...
           final var response = ((ResponseFailureException) cause).getResponse();
+
+          Tags tags;
+          if (response == null) {
+            tags = this.noResponseResourceTags;
+          } else {
+            tags = this.consumerVerticleContext.getTags().and(
+              Tag.of(Metrics.Tags.RESPONSE_CODE_CLASS, response.statusCode() / 100 + "xx"),
+              Tag.of(Metrics.Tags.RESPONSE_CODE, Integer.toString(response.statusCode())));
+          }
```
Comment (pierDipi) [inline data-plane/dispatcher/src/main/java/dev/knative/eventing/kafka/broker/dispatcher/impl/http/WebClientCloudEventSender.java:162]: We only need tags when the `if` at line 164 is true, so we should do this in that block

File: data-plane/core/src/main/java/dev/knative/eventing/kafka/broker/core/metrics/Metrics.java [hunk 7]
```diff
@@ -308,13 +314,15 @@ public static Counter.Builder eventCount(final io.micrometer.core.instrument.Tag
       .baseUnit(Metrics.Units.DIMENSIONLESS);
   }
 
-  public static Collection<Meter> searchResourceMeters(final MeterRegistry registry, final DataPlaneContract.Reference ref) {
+  public static Collection<Meter> searchResourceMeters(final MeterRegistry registry,
+      final DataPlaneContract.Reference ref) {
```
[the remark this answers is not in the document]
Comment (pierDipi) [reply data-plane/core/src/main/java/dev/knative/eventing/kafka/broker/core/metrics/Metrics.java:318]: Also the other method has the indentation off, see line 383 for an example

**Comment by pierDipi** [comment]
Closing this one as it doesn't seem there is interest in continuing it, we can keep it as reference

**Pull request closed without merging**